File: javascript/hamming/hamming.js
function hammingDifference(s1, s2) {
    if (!s1 || !s2 || s1.length != s2.length) {
        throw new Error('DNA strands must be of equal length.');
    }
    return s1.split('').reduce(function (diff, char, i) {
        return diff + (char === s2[i] ? 0 : 1);
    }, 0);
}

module.exports = {
    compute: hammingDifference
};

File: elixir/hamming/lib/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance(~c"AAGTCATA", ~c"TAGCGATC")
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) == length(strand2) do
    {:ok,
     Enum.zip(strand1, strand2)
     |> Enum.reduce(0, fn {a, b}, acc ->
       if a != b, do: acc + 1, else: acc
     end)}
  end

  def hamming_distance(_, _), do: {:error, "strands must be of equal length"}
end
